data/src/main/java/com/linkedin/data/template/WrappingArrayTemplate.java

```diff
@@ -43,15 +43,14 @@ protected WrappingArrayTemplate(DataList list, ArrayDataSchema schema, Class<E>
       throws TemplateOutputCastException
   {
     super(list, schema, elementClass, DataList.class);
-    _constructor = DataTemplateUtil.templateConstructor(elementClass, schema.getItems());
-    _cache = new DataObjectToObjectCache<E>(data().size());
+    _cache = new DataObjectToObjectCache<>(DataMapBuilder.getOptimumHashMapCapacityFromSize(list.size()));
   }
 
   @Override
   public boolean add(E element) throws ClassCastException
   {
     Object unwrapped;
-    boolean result = _list.add(unwrapped = unwrap(element));
+    boolean result = CheckedUtil.addWithoutChecking(_list, unwrapped = unwrap(element));
```

Comment:
This might not be safe? 
Atleast for existing record templates, we need to do the checks to ensure there are no cycles in the data.

Reply:
I know it is not safe. But someone inserting a datatemplate with cycles is shooting themselves in the foot, and should actually be penalized for it?

If you really have concerns about punishing them with StackOverflowError, one place to add checks would be Data.traverse where we can maintain an identitySet to track cycles and error out when serializing.

Reply:
Added check to Data.traverse as discussed offline.